src/parser/command.rs

```diff
@@ -223,6 +228,42 @@ pub fn parse_defframe<'a>(input: ParserInput<'a>) -> ParserResult<'a, Instructio
     ))
 }
 
+/// Parse the contents of a `DEFGATE` instruction.
+pub fn parse_defgate<'a>(input: ParserInput<'a>) -> ParserResult<'a, Instruction> {
+    let (input, name) = token!(Identifier(v))(input)?;
+    let (input, parameters) = opt(delimited(
+        token!(LParenthesis),
+        separated_list1(token!(Comma), token!(Variable(v))),
+        token!(RParenthesis),
+    ))(input)?;
+    let (input, gate_type) = opt(preceded(
+        token!(As),
+        alt((
+            map(token!(Matrix), |()| GateType::Matrix),
+            map(token!(Permutation), |()| GateType::Permutation),
+        )),
+    ))(input)?;
+    let gate_type = gate_type.unwrap_or(GateType::Matrix);
```

Comment:
Think `unwrap_or_else` is generally preferred

Reply:
I'd have to provide a function to `unwrap_or_else`, even though I need to return a constant value. That would be more verbose for no gain.

Reply:
Correct. `unwrap_or_else` is preferred when there is a function call being made, so that the function call is evaluated lazily instead of eagerly.